### ISSUE TITLE
Pass original validation error as an optional parameter of FieldError

### ISSUE THREAD

**Issue by PupoSDC** [react-hook-form/react-hook-form#7797]
**Is your feature request related to a problem? Please describe.**

It's currently not possible to access the original error thrown by validation libraries. 
This makes it very inconvenient to plug and play translation libraries in the react eco system.

For example, Zod throws the following validation error:

```ts
ZodError: [
  {
    "code": "too_small",
    "minimum": 3,
    "type": "string",
    "inclusive": true,
    "message": "error.nameMin",
    "path": [
      "name"
    ]
  }
]
```

but only this part is returned from `zodResolver`: 

```ts
name: name,
message: "error.nameMin",
ref: input,
type: "too_small",
```

Information like `minimum: 3` is lost, so it's impossible to wire up a hook based solution like `react-18next` to create dynamic translations like:

```ts
t(error.message, { minimum: error.minimum }
```

**Describe the solution you'd like**

At the Zod core level, the solution I'd like is implemented [here](https://github.com/react-hook-form/react-hook-form/pull/7792). It creates a "slot" to allow for an arbitrary error specific to a library to be returned as part of `FieldError`

This would then allow us to rewrite `zodResolver` to return the original error as part of the FieldError, which in turn would allow us to write something like:

```tsx
export const FormError : FunctionComponent<{ error: FieldError }> = ({
  error,
}) => {
  const { t } = useTranslation('formErrors');
  return error ? <span>{t(error.message, error.validationError)}</span> : null;
}
```

**Describe alternatives you've considered**

The above code already works if you dont care for typescript. In this [Poc Repository](https://github.com/PupoSDC/poc-stupid-server/blob/main/pages/index.tsx) I achieve what I want by implementing my own `ZodResolver`, which is just a copy paste of the existing one, with 2 lines changed. So that was one option.

Another option would be to have the translations being done at the schema, like

```ts
z.object({
  name: string().min(3, t("error.message", { minimum: 3 })
})
``` 

But this implies one of 2 things, either we have validation schemas getting created inside the react render cycle (because `t` is hook provided function) or we risk using a stale `t` function from the non hook provider. 

**Additional context**

This idea is loosely based on this post: https://github.com/react-hook-form/react-hook-form/discussions/3808#discussioncomment-261851

**Comment by bluebill1049** [comment]
Thanks for the feature request. This would require both core library and resolvers to support such, in fact, we are supporting various schema validation libraries, and having a consistent output would be nice. The first step would be this feature required: https://github.com/react-hook-form/react-hook-form/issues/3950

**Comment by bluebill1049** [comment]
perhaps to keep consistency

```
ZodError: [
  {
    "code": "too_small",
    "minimum": 3,
    "type": "string",
    "inclusive": true,
    "message": "error.nameMin",
    "path": [
      "name"
    ]
  }
]
```

to 

```
{
  expected: 5,
  received: xx,
}
```

**Comment by Moshyfawn** [comment]
I might be oversimplifying and not thinking it though, but what if we have a `nativeError` field along side the ones [@hookform/resolvers](https://github.com/react-hook-form/resolvers) already returns with the actual validation lib error object returned? Kinda like React has `event.nativeEvent` on the `event` object

**Comment by PupoSDC** [comment]
that is exactly what I propose! 😄 
I just called it [`validationError`](https://github.com/react-hook-form/react-hook-form/pull/7792/files#diff-070b8e435c80a2965d5d5bcd7df56033f7ecd901f0bb6bd44b4a581c41d70a44R18) instead of `nativeError`.

**Comment by PupoSDC** [comment]
I guess in my initial draft, I didn't consider that. I was just planning on leaving it as `unknown`. the user of the functionality would then have the responsability of casting the error:

```ts
const { validationError } = error as FieldError<ZodError> ;
```

But we could "drill" this type all the way down from the zod resolver signature. 

```ts
// This is for the Zod Resolver
// https://github.com/react-hook-form/resolvers/blob/master/zod/src/types.ts#L9
export type Resolver = <T extends z.Schema<any, any>>(
  schema: T,
  schemaOptions?: Partial<z.ParseParams>,
  factoryOptions?: { mode?: 'async' | 'sync' },
) => <TFieldValues extends FieldValues, TContext>(
  values: UnpackNestedValue<TFieldValues>,
  context: TContext | undefined,
  options: ResolverOptions<TFieldValues>,
) => Promise<ResolverResult<TFieldValues, ZodError>>;

// https://github.com/react-hook-form/react-hook-form/blob/master/src/types/resolvers.ts#L15
export type ResolverResult<TFieldValues extends FieldValues = FieldValues, ValidationError = unknown> =
  | ResolverSuccess<TFieldValues>
  | ResolverError<TFieldValues, ValidationError>;

// https://github.com/react-hook-form/react-hook-form/blob/master/src/types/resolvers.ts#L10
export type ResolverError<TFieldValues extends FieldValues = FieldValues, ValidationError = unknown> = = {
  values: {};
  errors: FieldErrors<TFieldValues, ValidationError>;
};

// https://github.com/react-hook-form/react-hook-form/blob/master/src/types/errors.ts#L26
export type FieldErrors<TFieldValues extends FieldValues = FieldValues, ValidationError = unknown> =
  DeepMap<DeepPartial<TFieldValues>, FieldError<ValidationError>>;

// https://github.com/react-hook-form/react-hook-form/blob/master/src/types/errors.ts#L26
export type FieldErrors<TFieldValues extends FieldValues = FieldValues, ValidationError = unknown> =
  DeepMap<DeepPartial<TFieldValues>, FieldError<ValidationError>>;

// https://github.com/react-hook-form/react-hook-form/blob/master/src/types/errors.ts#L13
export type FieldError<ValidationError = never> = {
  type: LiteralUnion<keyof RegisterOptions, string>;
  ref?: Ref;
  types?: MultipleFieldErrors;
  message?: Message;
  validationError?: ValidationError
};
```

Im not sure if RHF would pick this up though, would have to test it.

**Comment by Moshyfawn** [comment]
> how could make validationError type-safe with all the other validation schemas

I haven't looked close into each validation lib, but don't they expose their error type to slap onto the `validationError`?

Tho, to be fair, looking at this example of the wanted behavior
```js
t(error.message, { minimum: error.minimum }
```
Can you use your schema validation shape to do exactly that?
```js
const schema = z.object({
  name: z.string({ required_error: "Issue" }).min(3, "Length")
});
// will yield **3**, the min value you want to pass down to your `t` function
console.log(schema.shape.name.minLength)

t(error.message, { minimum: schema.shape.name.minLength }
```

**Comment by PupoSDC** [comment]
That is a solution we haven't though off. We might use it to get ourselves of the hole we are right now. 

However it lacks some elegancy, as that will force us to pass down the schema, or part of it, manually for each input, which removes away some of the abstraction RHF offers.

**Comment by Moshyfawn** [comment]
> However it lacks some elegancy, as that will force us to pass down the schema

I see your point. As a workaround, you create a Context to avoid passing down the schema manually as, I'd think, your schema is mostly static making use of `.refine()`, so it won't create a performance overhead.
But, yeah, it's no elegant abstraction and we should continue to think about a better solution

**Comment by bluebill1049** [comment]
What's the most essential information which is missing? we can go from there. 

https://github.com/react-hook-form/react-hook-form/issues/7797#issuecomment-1036010589

would my purpose solve most of your issues? I think i can align with build in validation as well.

**Comment by Moshyfawn** [comment]
> What's the most essential information which is missing? we can go from there.
> 
> [#7797 (comment)](https://github.com/react-hook-form/react-hook-form/issues/7797#issuecomment-1036010589)
> 
> would my purpose solve most of your issues? I think i can align with build in validation as well.

I'd say, it looks good. But I haven't seen other libs error info that can be useful

**Comment by bluebill1049** [comment]
If we can work out the fixed API, then we can work towards a resolver response as well. That way it would be type-safe and consistent as well.

**Comment by PupoSDC** [comment]
Your option would indeed get us out of the situation of the minimalist example, and I have nothing against it. 
I understand the appeal of having of abstracting away the different errors from different validators onto a common interface.

That being said I think providing us with a way to interact with the "native" error would be beneficial.  If not for anything else, `Zod` (And yup), provide ways to create [custom validators](https://github.com/colinhacks/zod#refine) with arbitrary parameters being returned for the error message.  And we seem to be a fan of those in the office ;) So it would just make it more future proof to have a way for us to untangle ourselves.

**Comment by balzdur** [comment]
I agree with @PupoSDC, the proposition won't cover every specific needs.

If the main concerns is to provide a common interface agnostic from any validation library, I think we need to focus on the initial need : provide all the parameters to `t` translation function. An "ideal" enhanced RHF error type would allow developer to pass something that looks like the `t` parameters interface. Unfortunately, it just move the issue : instead of defining an custom interface too close from a validation library; this define an interface too close from a translation library...

**Comment by bluebill1049** [comment]
> I agree with @PupoSDC, the proposition won't cover every specific needs.

Can we list some?

**Comment by Moshyfawn** [comment]
> > I agree with @PupoSDC, the proposition won't cover every specific needs.
> 
> Can we list some?

It seems like we're talking in general about some potential use-case we haven't come yet.
I'd say, let's implement the 
```js
{
  expected: 5,
  received: xx,
}
```
as it's the most useful info for error handling missing from whatever the field of possibilities with these validation libs. It solves the case where we need to have validation params for i18n, which was the initial request.
Next time the question comes up with some additional field needed from error object of some lib, we can have a look at how necessary it is to provide the full object.

Keep in mind that we don't wanna bloat the library with code that would further couple it with some tech. The resolvers solution was designed to be a common interface to integrate with whichever validation lib you want, not to make use any specific validation feature from a specific lib. Just the way [date-io](https://github.com/dmtrKovalenko/date-io) abstraction is.
See [this point](https://github.com/react-hook-form/react-hook-form/issues/7730#issuecomment-1028858309) on the matter as well

**Comment by PupoSDC** [comment]
The most obvious one, and the only one I can offer without starting to get creative, is indeed custom validators: 

```ts
const formSchema : z.ZodSchema<FormSchemaType> = z.object({
  name: z.string().min(3, "nameToShort").refine((val) => !val.includes("potato"), {
    message: "errors.bannedWord",
    params: {
      bannedWord: "potato",
    }
  })
})
```

This is just a dumb example based on one of my schemas, but I want to use `bannedWord` as a parameter for the translationKey `errors.BannedWord`.

```json
"errors": {
   "bannedWord": "Hey, you can't say {{ bannedWord }} here!"
}
```

The original Zod error is  (currently) the  only way for me to get access to this information since `name.value` is not necessarily just `potato`, could be `mr. potato`.

**Comment by balzdur** [comment]
I agree with you about the abstraction requirement @Moshyfawn but for that purpose, what about doing some inversion of control and let consumers of the library customise (at least a little bit) `type Message`  (something like I described [here](https://github.com/react-hook-form/react-hook-form/discussions/3808#discussioncomment-2150621))

I'm not an expert of the internal of RHF but I presume every logic based around form state errors is built over error being defined or not. If the internal structure of the error is not something important we could consider to bring some genericity on it. 

Again those are ideas based on my personal feeling of the lib as a consumer (not a core maintainer) and that raises issues around the type safety of those errors. 

Another aspect that might biased my perception is the way I use the lib. We defined some generic components based on [this idea from the doc](https://react-hook-form.com/advanced-usage/#SmartFormComponent) and thus, we try to bundle in generic component the error message display. This means :
- retrieving error message (from `useControler`)
- translate it 
- display with some UI components

The translation part is the bottleneck. For now, we translate it at the schema level so we do not need to do it internally in the component implementation. My personal needs for this feature is to address it internally (for the reason exposed both in this Issue + the linked discussion).

**Comment by Moshyfawn** [comment]
Don't get me wrong, I totally see your point. Having the same `event.nativeEvent` only for RHF for validation schemas would mean no more "expose _this_ field from _this_ validation lib", which would make for great DX and more tech integration opportunities. At least, on paper.

I'm more concerned about the potential coupling and type-safety issues + the way we talk about how unified error output _(the solution @bluebill1049 proposed)_ wouldn't cover all potential use-cases; ATM it sounds like a very blown out feature request. The proposed solution covers the use-case described. The idea behind what I was saying previously is to potentially iterate on the solution, but to start with having the "missing" fields add, currently preventing from moving forward with the described feature.

To sum up, there's a scope RHF / @hookform/resolvers have and we outta closely consider every piece of code we take in to maintain in the future. I'm not trying to dismiss your request, but rather pushing for having some solution built that would improve the situation for the use-case, while we're considering out next steps / discussion other potential solutions to apply in the future.

**Comment by PupoSDC** [comment]
Ok, I understand your vision, i just sadly disagree with you. Something that might have been lost in translation is that my original request was not to suport `t(error.message, { minimum: error.minimum })`. That was just the most minimalist of examples I was able to find to explain the use case. 

Most of our schemas are a bit more complex than that and rely quite a bit on `refine`. + custom messages / parameters. So the example of  [this message](https://github.com/react-hook-form/react-hook-form/issues/7797#issuecomment-1036202193). Reflects a lot better the needs of our team.

**Comment by bluebill1049** [comment]
Thanks for all the feedback. I understand where are your concern coming from. However, there is a general direction We need to take consider very carefully for the library not only to support or in favor of any schema validation library (there is build-in as well), so whatever decision we make need to fulfill both categories. This is just like another tradeoff we are making every day in software development. One of the reasons the error messages need to have a proper interface is that we need to set up the standard and why we are building this library in the first place. Eg, if you are switching schema validation libraries and nothing should break. If you are using typescript, then if we can have some sort of type guard would be nice but in this case, I would not die for the type-safe, but at least the shape is important. So just like @Moshyfawn mentioned, our goal is to try to solve the most important missing piece first, then progressively enhance the requirement, hopefully, it will cover most use cases. I understand this may not be the desired outcome that you want, but I believe this is a better approach to this problem.

So what's next? We will wait for the feature required to see more upvotes, then start to introduce the following two attributes to the errors object: `received` and `expected`, if there are more common usage, we can consider introducing more error interface. 

```diff
type FieldError = {
  type: LiteralUnion<keyof RegisterOptions, string>;
  ref?: Ref;
  types?: MultipleFieldErrors;
  message?: Message;
+ received?: unknown;
+ expected?: unknown;
}
```
